### DISCUSSION TITLE
[Question] ASR finetuning is in progress, but the train loss is no longer decreasing.

### DISCUSSION THREAD

**Discussion by jyseol-inno** [NVIDIA-NeMo/NeMo#4492] [Q&A]
Hello.
Referring to the [https://colab.research.google.com/github/NVIDIA/NeMo/blob/main/tutorials/asr/ASR_CTC_Language_Finetuning.ipynb#scrollTo=TSTb6b5DriWG](https://colab.research.google.com/github/NVIDIA/NeMo/blob/main/tutorials/asr/ASR_CTC_Language_Finetuning.ipynb#scrollTo=TSTb6b5DriWG) , i use 3,000 hours of Korean voice data, and the subword tokenizer uses bpe to make a vocab of 10000 size.

5epoch is currently in progress, and the value of train loss from 2epoch is
It doesn't go down any further between 50 and 65. What should we look at in a situation like this?

The code I wrote is as follows.

```python
# NeMo's ASR collection - this collections contains complete ASR models and
# building blocks (modules) for ASR
import nemo
from nemo.collections.asr.models import ASRModel
import pytorch_lightning as pl
from omegaconf import DictConfig, OmegaConf, open_dict
import copy
from nemo.utils import logging, exp_manager
from tqdm.auto import tqdm

# Load the model with an encoder that supports Adapters
model = ASRModel.from_pretrained('stt_en_citrinet_256', map_location='cpu')

# Preserve the decoder parameters in case weight matching can be done later
pretrained_decoder = model.decoder.state_dict()

# Insert preserved model weights if shapes match
if model.decoder.decoder_layers[0].weight.shape == pretrained_decoder['decoder_layers.0.weight'].shape:
    model.decoder.load_state_dict(pretrained_decoder)
    logging.info("Decoder shapes matched - restored weights from pre-trained model")
else:
    logging.info("\nDecoder shapes did not match - could not restore decoder weights from pre-trained model.")

#Similar to the Character-based models, we can freeze the encoder and unfreeze the batch normalization layers if the dataset is tiny.

freeze_encoder = False

if freeze_encoder:
  model.encoder.freeze()
  model.encoder.apply(enable_bn_se)
  logging.info("Model encoder has been frozen")
else:
  model.encoder.unfreeze()
  logging.info("Model encoder has been un-frozen")

#Update config
cfg = copy.deepcopy(model.cfg)

print(OmegaConf.to_yaml(cfg.train_ds))

model.cfg['tokenizer']['dir'] = "./tokenizers/kconfspeech/tokenizer_spe_bpe_v10000/"  # note this is a directory, not a path to a vocabulary file
model.cfg['tokenizer']['type'] = "bpe"

# Change vocab
model.change_vocabulary(new_tokenizer_dir = model.cfg['tokenizer']['dir'],
                        new_tokenizer_type = model.cfg['tokenizer']['type'])

# Update paths to dataset
model.cfg['train_ds']['manifest_filepath'] = './train.json'
model.cfg['train_ds']['batch_size'] = 32
#model.cfg['train_ds']['use_start_end_token'] = True
model.cfg['validation_ds']['manifest_filepath'] = './valid.json'
model.cfg['validation_ds']['batch_size'] = 32
#model.cfg['validation_ds']['use_start_end_token'] = True
model.setup_training_data(train_data_config=model.cfg['train_ds'])
model.setup_validation_data(val_data_config=model.cfg['validation_ds'])

print(OmegaConf.to_yaml(cfg.optim))

with open_dict(model.cfg.optim):
  model.cfg.optim.lr = 0.025
  model.cfg.optim.weight_decay = 0.001
  model.cfg.optim.sched.warmup_steps = None  # Remove default number of steps of warmup
  model.cfg.optim.sched.warmup_ratio = 0.10  # 10 % warmup
  model.cfg.optim.sched.min_lr = 1e-9

with open_dict(model.cfg.spec_augment):
  model.cfg.spec_augment.freq_masks = 2
  model.cfg.spec_augment.freq_width = 25
  model.cfg.spec_augment.time_masks = 10
  model.cfg.spec_augment.time_width = 0.05

model.spec_augmentation = model.from_config_dict(model.cfg.spec_augment)

use_cer = True
log_prediction = True
model._wer.use_cer = use_cer
model._wer.log_prediction = log_prediction

EPOCHS = 100  # 100 epochs would provide better results

trainer = pl.Trainer(devices=3,
                      strategy="ddp",
                      accelerator='gpu',
                      max_epochs=EPOCHS,
                      enable_checkpointing=False,
                      logger=False,
                      log_every_n_steps=100,
                      check_val_every_n_epoch=10)

# Setup model with the trainer
model.set_trainer(trainer)

# finally, update the model's internal config
model.cfg = model._cfg

# Environment variable generally used for multi-node multi-gpu training.
# In notebook environments, this flag is unnecessary and can cause logs of multiple training runs to overwrite each other.
#os.environ.pop('NEMO_EXPM_VERSION', None)
LANGUAGE = 'kor'
config = exp_manager.ExpManagerConfig(
    exp_dir=f'experiments/lang-{LANGUAGE}/',
    name=f"ASR-Model-Language-{LANGUAGE}",
    checkpoint_callback_params=exp_manager.CallbackParams(
        monitor="val_wer",
        mode="min",
        always_save_nemo=True,
        save_best_model=True,
    ),
)

config = OmegaConf.structured(config)

logdir = exp_manager.exp_manager(trainer, config)

trainer.fit(model)

save_path = f"Model-{LANGUAGE}.nemo"
model.save_to(f"{save_path}")

```

my `model.cfg`:
```bash
cfg:
  sample_rate: 16000
  train_ds:
    manifest_filepath: ./train.json
    sample_rate: 16000
    batch_size: 32
    trim_silence: true
    max_duration: 16.7
    shuffle: true
    is_tarred: false
    tarred_audio_filepaths: null
    use_start_end_token: false
  validation_ds:
    manifest_filepath: ./valid.json
    sample_rate: 16000
    batch_size: 32
    shuffle: false
    use_start_end_token: false
  test_ds:
    manifest_filepath: null
    sample_rate: 16000
    batch_size: 32
    shuffle: false
    use_start_end_token: false
  model_defaults:
    repeat: 5
    dropout: 0.0
    separable: true
    se: true
    se_context_size: -1
  tokenizer:
    dir: ./tokenizers/kconfspeech/tokenizer_spe_bpe_v10000/
    type: bpe
    model_path: /app/nemo/tokenizers/kconfspeech/tokenizer_spe_bpe_v10000/tokenizer.model
    vocab_path: /app/nemo/tokenizers/kconfspeech/tokenizer_spe_bpe_v10000/vocab.txt
    spe_tokenizer_vocab: /app/nemo/tokenizers/kconfspeech/tokenizer_spe_bpe_v10000/tokenizer.vocab
  preprocessor:
    _target_: nemo.collections.asr.modules.AudioToMelSpectrogramPreprocessor
    sample_rate: 16000
    normalize: per_feature
    window_size: 0.025
    window_stride: 0.01
    window: hann
    features: 80
    n_fft: 512
    frame_splicing: 1
    dither: 1.0e-05
    pad_to: 16
    stft_conv: false
  spec_augment:
    _target_: nemo.collections.asr.modules.SpectrogramAugmentation
    freq_masks: 2
    time_masks: 10
    freq_width: 25
    time_width: 0.05
  optim:
    name: novograd
    lr: 0.025
    betas:
    - 0.8
    - 0.25
    weight_decay: 0.001
    sched:
      name: CosineAnnealing
      warmup_steps: null
      warmup_ratio: 0.1
      min_lr: 1.0e-09
      last_epoch: -1
                                                      
```

`tesnsorboard train loss`:
![image](https://user-images.githubusercontent.com/100547581/177021195-dc8f9976-fe70-4fd1-bc35-3a56325c6a43.png)

**Comment by titu1994** [comment]
Some notes - you should do the "Insert preserved model weights if shapes" part *after you change the vocabulary* - since changing the vocabulary is what changes the weights of the decoder. 

I also don't think you should use Citrinet for Korean language, the vocabulary is vast and Conformer can handle it much better - you can use Conformer small for experiments, nearly same size as the Citrinet one. 

Now as to why loss doesn't decrease further - two things 

- CTC models take a long time to converge - 5 epochs is not enough (even with a pretrained encoder+decoder, and here you can't load the decoder since your vocab size is so much larger than the original models)
- All Nemo ASR model have Spec Augment in them active by default - the small model uses 2 time masks I believe, which is recommended - you are overriding with 10 time masks which is too much for such a small model. 2 time nasks is small enough such that you will have a small amount of train loss always present (it depends on dataset, for example with LS it's 20-30, for some datasets it's 5-10 and others it's 40-60). You can try overfitting your model by setting spec aug to zero but I won't recommend that. 

If you switch to conformer - go to the config page and copy paste it's optim section - and reduce LR to 0.2 or 0.5 (note that for Conformer + Noam, LS is a multiplier not an actual LR value). 

Another note, since you're using a small model, but your dataset is so large, you may really want to use larger models - but that depends on your compute available. Just know that small models have their limits and will eventually not be able to learn on more data (for example, our small models don't learn much beyond 10k hours of speech). 

And finally, if you want the best possible wer, we suggest using Conformer Transducer small instead of Conformer CTC. Much faster speed of convergence, and usually much better wer.

**Reply by jyseol-inno** [reply]
@titu1994 
I appreciate your reply.
Is there a tutorial where I can use the Conformer CTC or Conformer Transducer model?

**Reply by titu1994** [reply]
No, there isn't a tutorial per model. You'll have to swap out the model name and check what it's config is like. We may write an RNNT finetuning guide in the future but right now there's 95% overlap with CTC finetuning so there's no need for it

**Reply by 1shershah** [reply]
> If you switch to conformer - go to the config page and copy paste it's optim section - and reduce LR to 0.2 or 0.5 (note that for Conformer + Noam, LS is a multiplier not an actual LR value).
> 

Any one have a idea what is titu1994 referring to as "If you switch to conformer - go to the config page and copy paste it's optim section" - which config page ?

**Reply by titu1994** [reply]
Examples config page. https://github.com/NVIDIA/NeMo/blob/main/examples/asr/conf/conformer/conformer_transducer_bpe.yaml#L198
